### PR TITLE
Don't return profiles in reference lists/starter packs that have a block relationship with the owner 

### PR DESCRIPTION
**Pull request by haileyok**

### CODE REVIEW

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 1]
```diff
@@ -68,9 +69,41 @@ const hydration = async (
...
+    // We show all users regardless of blocks if the viewer is the owner of the list, so no need to hydrate them
+    params.hydrateCtx.viewer !== ownerDid
+  ) {
+    const pairs = new Map()
+    for (const { did } of listitems) {
+      pairs.set(did, [did, ownerDid])
+    }
+    const bidirectionalBlocks =
+      await ctx.hydrator.hydrateBidirectionalBlocks(pairs)
+    return mergeStates(listState, { ...profileState, bidirectionalBlocks })
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:86]: If `profileState` contains some `bidirectionalBlocks` they would get clobbered here, so we'd be safer with something like `mergeManyStates(listState, profileState, { bidirectionalBlocks })`.

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 2]
```diff
@@ -68,9 +69,41 @@ const hydration = async (
...
+  const ownerDid = new AtUri(listUri).hostname
+  if (
+    list?.record.purpose === 'app.bsky.graph.defs#referencelist' &&
+    // We show all users regardless of blocks if the viewer is the owner of the list, so no need to hydrate them
+    params.hydrateCtx.viewer !== ownerDid
+  ) {
+    const pairs = new Map()
+    for (const { did } of listitems) {
+      pairs.set(did, [did, ownerDid])
+    }
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:83]: If I'm not mistaken we'd be good with a single item `ownerDid -> listitems.map((li) => li.did)`.

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 3]
```diff
@@ -68,9 +69,41 @@ const hydration = async (
...
+    const bidirectionalBlocks =
+      await ctx.hydrator.hydrateBidirectionalBlocks(pairs)
+    return mergeStates(listState, { ...profileState, bidirectionalBlocks })
+  }
+
   return mergeStates(listState, profileState)
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:89]: Would be kind of slick if we could remove the branch in here and end-up with something like:
```ts
const bidirectionalBlocks = maybeGetBlocksForReferenceList(/* args */)
return mergeManyStates(listState, profileState, { bidirectionalBlocks })
```

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 4]
```diff
@@ -68,9 +70,41 @@ const hydration = async (
...
 }
 
+const noBlocks = (input: RulesFnInput<Context, Params, SkeletonState>) => {
+  const { skeleton, hydration } = input
+  if (!hydration.bidirectionalBlocks) {
+    return skeleton
+  }
+  skeleton.listitems = skeleton.listitems.filter(({ did }) => {
+    const blocks = hydration.bidirectionalBlocks?.get(did)?.values() ?? []
+    for (const block of blocks || []) {
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:100]: Already got the default rockin above.
```suggestion
    for (const block of blocks) {
```

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 5]
```diff
@@ -68,7 +70,31 @@ const hydration = async (
...
+  if (!hydration.bidirectionalBlocks) {
+    return skeleton
+  }
+  skeleton.listitems = skeleton.listitems.filter(({ did }) => {
+    const blocks = hydration.bidirectionalBlocks?.get(did)?.values() || []
+    for (const block of blocks) {
+      if (block) return false
+    }
+    return true
+  })
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:96]: I very well may have this wrong, but I think the way hydration is working, we want something like this:
```suggestion
  const blocks = hydration.bidirectionalBlocks?.get(ownerDid)
  skeleton.listitems = skeleton.listitems.filter(({ did }) => {
    return blocks?.get(did)
  })
```

Comment (devinivy) [reply packages/bsky/src/api/app/bsky/graph/getList.ts:96]: (I believe `bidirectionalBlocks` is a map of maps `DID -> DID -> boolean`.)

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 6]
```diff
@@ -88,6 +114,31 @@ const presentation = (
   return { list, items, cursor }
 }
 
+const maybeGetBlocksForReferenceList = async (input: {
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:117]: We may need similar logic in `app.bsky.graph.getStarterPack` for `starterPack.listItemsSample`.

File: packages/bsky/src/api/app/bsky/graph/getList.ts (hunk 5 above)
Comment (haileyok) [reply packages/bsky/src/api/app/bsky/graph/getList.ts:96]: hmm let me try that again. i think i actually tried that but now curious again.

Comment (haileyok) [reply packages/bsky/src/api/app/bsky/graph/getList.ts:96]: argh, i wasn't building before test lol, sick.

File: packages/bsky/src/api/app/bsky/graph/getList.ts (hunk 6 above)
Comment (haileyok) [reply packages/bsky/src/api/app/bsky/graph/getList.ts:115]: Added something for that, tell me if it makes sense or not - seemed to me to make the most sense to handle it inside of `hydrateStarterPacks()`.

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 7]
```diff
@@ -68,7 +70,29 @@ const hydration = async (
...
+  const bidirectionalBlocks = await maybeGetBlocksForReferenceList({
+    ctx,
+    params,
+    skeleton,
+    listState,
+  })
+  return [listState, profileState, { bidirectionalBlocks }].reduce(
+    mergeStates,
+    {} as HydrationState,
+  )
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:82]: The `mergeManyStates()` helper will reduce for you!  May need to export it, but it lives right next to `mergeStates()`.

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 8]
```diff
@@ -68,7 +70,29 @@ const hydration = async (
...
+}
+
+const noBlocks = (input: RulesFnInput<Context, Params, SkeletonState>) => {
+  const { skeleton, hydration } = input
+  if (!hydration.bidirectionalBlocks) {
+    return skeleton
+  }
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:89]: This branch is fine, but IMO also cool to nix it.  If we want to keep it, can get rid of the defensive check below `hydration.bidirectionalBlocks?.get(creator)`.

File: packages/bsky/src/api/app/bsky/graph/getList.ts [hunk 9]
```diff
@@ -68,7 +70,29 @@ const hydration = async (
...
+  const creator = new AtUri(skeleton.listUri).hostname
```
Comment (devinivy) [inline packages/bsky/src/api/app/bsky/graph/getList.ts:90]: There's a helper `didFromUri()` that could be used here.

File: packages/bsky/src/hydration/hydrator.ts [hunk 10]
```diff
@@ -650,6 +651,19 @@ export class Hydrator {
     const listMemberDids = listsMembers.flatMap((lm) =>
       lm.listitems.map((li) => li.did),
     )
+    const listBlocks: Map<string, Blocks> = new Map()
+    await Promise.all(
+      listUris.map(async (uri) => {
+        const creator = new AtUri(uri).hostname
```
Comment (devinivy) [inline packages/bsky/src/hydration/hydrator.ts:657]: Suggest using `didFromUri()` here.

Comment (devinivy) [reply packages/bsky/src/hydration/hydrator.ts:657]: (There are a couple more of these lying around but wont leave comments on all of them.)

File: packages/bsky/src/hydration/hydrator.ts [hunk 11]
```diff
@@ -650,6 +651,19 @@ export class Hydrator {
     const listMemberDids = listsMembers.flatMap((lm) =>
       lm.listitems.map((li) => li.did),
     )
+    const listBlocks: Map<string, Blocks> = new Map()
+    await Promise.all(
+      listUris.map(async (uri) => {
```
Comment (devinivy) [inline packages/bsky/src/hydration/hydrator.ts:656]: I'm not sure that breaking the blocks up by list helps us out a ton, can probably get rid of that idea.
```ts
const listCreatorMemberPairs = [...listMembersByList.entries()]
  .flatMap(([listUri, members]) => {
    const creator = didFromUri(listUri)
    return members.map((member) => [creator, member.did])
  })

const blocks = await this.graph.getBidirectionalBlocks(listCreatorMemberPairs)

// ...

const creator = didFromUri(sp.record.list)
// update aggregation with list items for top 12 most followed members
agg.listItemSampleUris = [...members.listitems]
  .filter((li) => !blocks.isBlocked(creator, li.did))
  // ...
```

**Review by devinivy — APPROVED** [review]
Good stuff 👍

**Pull request merged**